### ISSUE TITLE
Amaranth conflicts: pinned package won't install; later ones fail to produce bitstream

### ISSUE THREAD

**Issue by zzattack** [courk/esp-cpa#3]
Installing the pinned version of amaranth fails with a `TypeError: parse() missing 1  required positional argument` when using `poetry install` to do so:

> 
>   ChefBuildError
> 
>   Backend subprocess exited when trying to invoke get_requires_for_build_wheel
>   
>   Traceback (most recent call last):
>     File "/root/.local/pipx/venvs/poetry/lib/python3.11/site-packages/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
>       main()
>     File "/root/.local/pipx/venvs/poetry/lib/python3.11/site-packages/pyproject_hooks/_in_process/_in_process.py", line 335, in main
>       json_out['return_val'] = hook(**hook_input['kwargs'])
>                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     File "/root/.local/pipx/venvs/poetry/lib/python3.11/site-packages/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
>       return hook(config_settings)
>              ^^^^^^^^^^^^^^^^^^^^^
>     File "/tmp/tmpbp4myq01/.venv/lib/python3.11/site-packages/setuptools/build_meta.py", line 341, in get_requires_for_build_wheel
>       return self._get_build_requires(config_settings, requirements=['wheel'])
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     File "/tmp/tmpbp4myq01/.venv/lib/python3.11/site-packages/setuptools/build_meta.py", line 323, in _get_build_requires
>       self.run_setup()
>     File "/tmp/tmpbp4myq01/.venv/lib/python3.11/site-packages/setuptools/build_meta.py", line 338, in run_setup
>       exec(code, locals())
>     File "<string>", line 21, in <module>
>     File "<string>", line 7, in doc_version
>   TypeError: parse() missing 1 required positional argument: 'config'
>   
> 
>   at ~/.local/pipx/venvs/poetry/lib/python3.11/site-packages/poetry/installation/chef.py:164 in _prepare
>       160│ 
>       161│                 error = ChefBuildError("\n\n".join(message_parts))
>       162│ 
>       163│             if error is not None:
>     → 164│                 raise error from None
>       165│ 
>       166│             return path
>       167│ 
>       168│     def _prepare_sdist(self, archive: Path, destination: Path | None = None) -> Path:
> 

Amaranth changed build system to PDM, official releases >= 0.4.0 do install successfully.

Unfortunately, the `poetry run ctrl configure-board`  then fails with:
`TypeError: Object flipped((rec qspi_clk_0 i)) cannot be converted to an Amaranth value` .

**Comment by courk** [comment]
Hello,

Initially, I couldn’t replicate the `poetry` problem on my machine. I performed a clean clone of the repo and then ran `poetry install`, and everything went smoothly.

However, when I tried it on a laptop I hadn’t used for this project before, I encountered the same issue. It seems there might be something odd with the poetry cache that allows it to work on my primary machine.

In any case, updating to an actual release of _Amaranth_ instead of pinning to a random commit (which was likely the main branch head when I began working on this project) is a good idea.

After updating to a more recent release of _Amaranth_, I initially encountered the error message you reported, and it seems to be related to [recent changes between v0.3 and v0.4](https://whitequark.github.io/amaranth/docs/rfc-2-document/changes.html#migrating-from-version-0-3).

Applying this simple patch seems to be enough for things to work. 

```diff
diff --git a/esp_cpa_board/gateware/top.py b/esp_cpa_board/gateware/top.py
index 8f63587..c8b2bdc 100755
--- a/esp_cpa_board/gateware/top.py
+++ b/esp_cpa_board/gateware/top.py
@@ -65,9 +65,9 @@ class Top(Elaboratable):

dut_clk = platform.request("dut_clk", 0)

- qspi_cs = platform.request("qspi_cs", 0)
+ qspi_cs = platform.request("qspi_cs", 0).i
p_qspi_cs = platform.request("p_qspi_cs", 0)
- qspi_clk = platform.request("qspi_clk", 0)
+ qspi_clk = platform.request("qspi_clk", 0).i
qspi_in = platform.request("qspi_sd1", 0)

heat_ctrl = platform.request("heat_ctrl", 0)
```


However, note that I didn’t test the resulting bitstreams against the actual hardware yet. I don’t have an assembled _esp-cpa_ at my immediate disposal anymore; I shipped all my hardware to Espressif’s headquarters.
